pkg/controller/job/job_controller.go

```diff
@@ -646,32 +724,240 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+// when Job tracking with finalizers is disabled.
+func (jm *Controller) removeTrackingFinalizersFromAllPods(pods []*v1.Pod) error {
+	var podsWithFinalizer []*v1.Pod
+	for _, pod := range pods {
+		if hasJobTrackingFinalizer(pod) {
+			podsWithFinalizer = append(podsWithFinalizer, pod)
+		}
+	}
+	if len(podsWithFinalizer) == 0 {
+		return nil
+	}
+	_, err := jm.removeTrackingFinalizerFromPods(podsWithFinalizer)
+	return err
+}
+
+// trackJobStatusAndRemoveFinalizers does:
+// 1. Add finished Pods to .status.uncountedTerminatedPods
+// 2. Remove the finalizers from the Pods if they completed or were removed
+//    or the job was removed.
+// 3. Increment job counters for pods that no longer have a finalizer.
+// 4. Add Complete condition if satisfied with current counters.
+// It does this in a controlled way such that the size of .status doesn't grow
+// too much.
+func (jm *Controller) trackJobStatusAndRemoveFinalizers(job *batch.Job, pods []*v1.Pod, succeededIndexes orderedIntervals, uncounted uncountedTerminatedPods, finishedCond *batch.JobCondition, needsFlush bool) error {
+	isIndexed := isIndexedJob(job)
+	var podsToRemoveFinalizer []*v1.Pod
+	uncountedStatus := job.Status.UncountedTerminatedPods
+	var newSucceededIndexes []int
+	flushUncountedAndRemoveFinalizers := func(last bool) error {
+		if len(newSucceededIndexes) > 0 {
+			needsFlush = true
+			succeededIndexes = succeededIndexes.withOrderedIndexes(newSucceededIndexes)
+			job.Status.Succeeded = int32(succeededIndexes.total())
+			job.Status.CompletedIndexes = succeededIndexes.String()
+		}
+		if needsFlush {
+			if err := jm.updateStatusHandler(job); err != nil {
+				return fmt.Errorf("adding uncounted pods to status: %w", err)
+			}
+			needsFlush = false
+		}
+		var rmErr error
+		if len(podsToRemoveFinalizer) > 0 {
+			podsToRemoveFinalizer, rmErr = jm.removeTrackingFinalizerFromPods(podsToRemoveFinalizer)
+		}
+		uidsWithFinalizer := make(sets.String, len(podsToRemoveFinalizer))
+		for _, p := range podsToRemoveFinalizer {
+			uidsWithFinalizer.Insert(string(p.UID))
+		}
+		newUncounted := updateUncountedWithFailedFinalizerRemovals(uncountedStatus.Succeeded, uidsWithFinalizer)
+		if len(newUncounted) != len(uncountedStatus.Succeeded) {
+			needsFlush = true
+			job.Status.Succeeded += int32(len(uncountedStatus.Succeeded) - len(newUncounted))
+			uncountedStatus.Succeeded = newUncounted
+		}
+		newUncounted = updateUncountedWithFailedFinalizerRemovals(uncountedStatus.Failed, uidsWithFinalizer)
+		if len(newUncounted) != len(uncountedStatus.Failed) {
+			needsFlush = true
+			job.Status.Failed += int32(len(uncountedStatus.Failed) - len(newUncounted))
+			uncountedStatus.Failed = newUncounted
+		}
+		if rmErr == nil && last && jm.enactJobFinished(job, finishedCond) {
+			needsFlush = true
+		}
+		if needsFlush && (last || rmErr != nil) {
+			if err := jm.updateStatusHandler(job); err != nil {
```

Comment:
As mentioned before, there should be clear inputs, reaction to it (this method) and status update and return. Currently there are possibly 2 status updates, which might be unnecessary.

Reply:
See the update. But to be clear, both status updates are necessary.